README.md
# coinvwap

![](assets/python_3.8.svg?raw=true "pylint")
![](assets/pylint_enabled.svg?raw=true "pylint")
![](assets/pylint_10_10.svg?raw=true "pylint")
![](assets/black_enabled.svg?raw=true "black")
![](assets/mkdocs_enabled.svg?raw=true "mkdocs")
![](assets/pytest_pass.svg?raw=true "pytest")
![](assets/coverage.svg?raw=true "coverage")

## Intro

Coinvwap retrieves a data feed from the coinbase websocket and subscribe to the matches channel.

It pulls data for the following three trading pairs:

- BTC-USD
- ETH-USD
- ETH-BTC

It calculates the [VWAP](https://en.wikipedia.org/wiki/Volume-weighted_average_price)
per trading pair using a sliding window of 200 data points.
Meaning, when a new data point arrives through the websocket feed the oldest data
point will fall off and the new one will be added such that no more than 200 data
points are included in the calculation.

- The first 200 updates will have less than 200 data points included.
- Stream the resulting VWAP values on each websocket update.
- Prints to stdout.

## Design

Coinvwap is written in pure Python without any dependencies, although
to speed up development it contains a class borrowed from [websocket-client library](https://github.com/websocket-client/websocket-client),
precisely `ABNF` class.

During development close follow was kept with:

- [RFC6455, especially #section-5.2](https://datatracker.ietf.org/doc/html/rfc6455#section-5.2)
- [RFC5234](https://datatracker.ietf.org/doc/html/rfc5234)

## Development

Requirements:

- Python >=3.8
- [Poetry](https://python-poetry.org/docs/#installation)

Running:

0. For quick run: `poetry run python src/main.py`

1. For dependencies: `poetry install`

2. For tests: `poetry run ./scripts/run_tests.sh`

3. For build: `poetry build`

---

## Customization

In case of a need to change channel or fields which are read from the
websocket stream there is a wide range of arguments available through
the main object:

```python
cvp = Coinvwap(
    url="wss://ws-feed.exchange.coinbase.com/",
    product_ids=["BTC-USD", "ETH-USD", "ETH-BTC"],
    channel="ticker",
    type_field="ticker"
    price_field="price",
    quantity_field="last_size",
)
```

The code has been tested for a limited combinations of these.
All were not thoroughly tested, so reach out if you run into problems.

---

## Notes / Justifications

### "you may not use any external libraries"

I took "you may not use any external libraries" in the description very-very seriously. The question came just when I run into binary wrapping and then parsing of frames
[rfc6455#section-5.2](https://datatracker.ietf.org/doc/html/rfc6455#section-5.2).

I though I will send a question and make sure this is not an overkill. I even suspected that it might be a part of the task - ability to byte-stream processing based on RFC specs.

The end game is - I have done this in pure python without any websocket library.

### Ambiguity of the proper data use

It wasn't clear on where to get 'quantity' exactly, there is no
field of this name on any of the channels.

Eric suggested to use channel "match" which websocket reports
as unknown. Maybe it requires an account ..

According to the docs: "The ticker channel provides real-time
price updates every time a match happens.
It batches updates in case of cascading matches,
greatly reducing bandwidth requirements."

It seems like the same functionality as `ticker` offers with less bandwidth.
There is no `quantity`, so I used `last_size` which seemed closest.
Anyway, I used what I used - constructor arguments below allow
to customize the subscription.

**FIXME: User fields selected - to be verified, by the code reviewer.**

### Thread wrapping and watchdog

Listener can be wrapped in a thread or separate process
(like I did with the tests/test_integration.py).
This can be done mainly not to block the main thread,
use the output stream as a base for websocket server,
perform watchdog operations or else.

Ofc watchdog operation can be also handled on the OS level.

### Output piping

To pipe the output the report_fn can be passed.

```python
# ...
cvp.connect()
cvp.listen(report_fn=print)
```

### Websocket protocol

My websocket protocol client implementation is very shallow,
the code will not handle too many variations, but it works for Coinbase case!

I did what I could considering the limited time available and "no external libraries" restriction.

Framing protocol specs: https://datatracker.ietf.org/doc/html/rfc6455#section-5.2

Just for a little explanation of what's going on, a subscription frame request example:

```python
raw_sub = b'\x81\xfe\x00\x87\x16\x98\xa4\xe7m\xba\xd0\x9ef\xfd\x86\xdd6\xba\xd7\x92t\xeb\xc7\x95\x7f\xfa\xc1\xc5:\xb8\x86\x97d\xf7\xc0\x92u\xec\xfb\x8er\xeb\x86\xdd6\xc3\x86\xa5B\xdb\x89\xb2E\xdc\x86\xba:\xb8\x86\x84~\xf9\xca\x89s\xf4\xd7\xc5,\xb8\xff\xc5~\xfd\xc5\x95b\xfa\xc1\x86b\xba\x88\xc7m\xba\xca\x86{\xfd\x86\xdd6\xba\xd0\x8eu\xf3\xc1\x954\xb4\x84\xc5f\xea\xcb\x83c\xfb\xd0\xb8\x7f\xfc\xd7\xc5,\xb8\xff\xc5T\xcc\xe7\xcaC\xcb\xe0\xc5:\xb8\x86\xa5B\xdb\x89\xa0T\xc8\x86\xbak\xc5\xd9'
print("{:03b}".format(raw_sub[0]))
# gives: '10000001'
```

Based on the binary websocket frame model:
```
fin: 1
rsv1: 0
rsv2: 0
rsv3: 0
opcode: 1 = %x1 means text frame
mask
length
```

To avoid covering a lot of cases I focused on the type of responses from Coinbase.

Noted that they all come unmasked and within a certain length limits, so I dropped support for diverted variations and focused completely on this case.

Other coinbase websocket protocol specifics - not checking:
- masking
- sequence count indicating each sent message, there is no verification if I didn't miss any (no account has been used, so free feed is not complete anyway)
- frame lengths above LENGTH_63


---

## Future improvements

- more docstrings and comments
- websocket stream mocking
- thread wrapper
- more code coverage towards 100%
- more unit tests


docs/changelog.md
Changelog
=========


(unreleased)
------------
- Readme update. [Maciej]
- Headers refactor. [Maciej]
- Readme update. [Maciej]
- Payload test. [Maciej]
- Integration test improved; [Maciej]
- More docstrings; [Maciej]
- README update; [Maciej]
- README update; [Maciej]
- README update; [Maciej]
- README update; [Maciej]
- README update; [Maciej]
- README update; [Maciej]
- Typo fix; changelog update; [Maciej]
- More tests; more README; more checks; more coverage. [Maciej]
- Refactor; pylint fixes; package build prep; pytest config. [Maciej]
- Poc. [Maciej]




scripts/run_tests.sh
#!/bin/bash

PYTEST_COMMAND='python -B -m pytest -vvv -rfs -p no:cacheprovider'
PYLINT_COMMAND='pylint src --fail-under=10 --output-format=json:.pylint_output.json,colorized'

run_tests() {
    $PYTEST_COMMAND $@
    if [ "$?" == "1" ]; then
        exit 1
    fi
    rm -f assets/coverage.svg
    coverage-badge > assets/coverage.svg
}

run_pylint() {
    $PYLINT_COMMAND $@
    if [ "$?" == "1" ]; then
        exit 1
    fi
}

run_mkdocs() {
    gitchangelog > docs/changelog.md
    mkdocs build
}

run_tests $@
run_pylint $@
run_mkdocs $@


src/tests/test_main.py
import main


def test_main():
    """
    Dummy import test
    """
    assert str(type(main)) == "<class 'module'>"


src/tests/test_vwap.py
from vwap import VWAPStore


def test_vwap_store():
    payloads = [
        {
            "type": "match",
            "product_id": "ETH-BTC",
            "price": 0.05478,
            "quantity": 0.072342,
        },
        {
            "type": "match",
            "product_id": "ETH-BTC",
            "price": 0.04478,
            "quantity": 0.042342,
        },
        {
            "type": "match",
            "product_id": "ETH-BTC",
            "price": 0.03478,
            "quantity": 0.042342,
        },
    ]

    store = VWAPStore(
        product_ids=["BTC-USD", "ETH-USD", "ETH-BTC"],
        price_field="price",
        quantity_field="quantity",
        type_field="match",
    )
    for payload in payloads:
        store.store(payload)

    assert (
        store.report()
        == "BTC-USD\t000.000000\nETH-USD\t000.000000\nETH-BTC\t000.046691\n"
    )


src/vwap.py
from collections import deque
from typing import Dict, List


class VWAPStore:
    """
    Storage for the 200 data points and computation of VWAP indicator.
    """

    def __init__(
        self,
        product_ids: List[str],
        price_field: str,
        quantity_field: str,
        type_field: str = "ticker",
    ) -> None:
        self.product_ids = product_ids
        self.price_field = price_field
        self.quantity_field = quantity_field
        self.prices_n_vols = {
            product_id: deque(maxlen=200) for product_id in product_ids
        }
        self.type_field = type_field

    def store(self, payload: Dict) -> None:
        if payload["type"] == self.type_field:
            self._put(
                payload["product_id"],
                float(payload[self.price_field]),
                float(payload[self.quantity_field]),
            )

    def _put(self, product_id: str, price: float, last_size: float) -> None:
        if (
            product_id
            in self.prices_n_vols.keys()  # pylint: disable=consider-iterating-dictionary
        ):
            self.prices_n_vols[product_id].append((price, last_size))
        else:
            # TODO: more pro exception handling to be added
            pass

    @staticmethod
    def vwap(prices: List[float], quantity: List[float]) -> float:
        if len(quantity) == 0:
            return 0.0
        return sum([pj * qj for pj, qj in zip(prices, quantity)]) / sum(quantity)

    @staticmethod
    def vwap_formated(prices: List[float], vols: List[float]):
        return f"{VWAPStore.vwap(prices, vols):010f}"

    def report(self, point_counts=False) -> str:
        buff = ""
        for product_id, product_item in self.prices_n_vols.items():
            prices = [val[0] for val in product_item]
            vols = [val[1] for val in product_item]
            buff += f"{product_id}\t{VWAPStore.vwap_formated(prices, vols)}"
            if point_counts:
                buff += f"\tpoints:\t{len(product_item)}"
            buff += "\n"
        return buff


class VWAPStoreFast:  # pylint: disable=too-many-instance-attributes
    """
    Storage for the 200 data points and computation of VWAP indicator.
    """

    def __init__(
        self,
        product_ids: List[str],
        price_field: str,
        quantity_field: str,
        type_field: str = "ticker",
    ) -> None:
        self.product_ids = product_ids
        self.price_field = price_field
        self.quantity_field = quantity_field
        self.prices_n_vols = {product_id: [] for product_id in product_ids}
        self.sum_of_pjqj = {product_id: 0.0 for product_id in product_ids}
        self.sum_of_qj = {product_id: 0.0 for product_id in product_ids}
        self.count_of_j = 0
        self.type_field = type_field

    def store(self, payload: Dict) -> None:
        if payload["type"] == self.type_field:
            self._put(
                payload["product_id"],
                float(payload[self.price_field]),
                float(payload[self.quantity_field]),
            )

    def _put(self, product_id: str, price: float, last_size: float) -> None:
        if self.count_of_j >= 200:
            # Remove the oldest
            self.sum_of_pjqj[product_id] -= (
                self.prices_n_vols[product_id][0][0]
                * self.prices_n_vols[product_id][0][1]
            )
            self.sum_of_qj[product_id] -= self.prices_n_vols[product_id][0][1]
            self.count_of_j -= 1
            del self.prices_n_vols[product_id][0]
        # Add the newest
        self.prices_n_vols[product_id].append((price, last_size))
        self.sum_of_pjqj[product_id] += price * last_size
        self.sum_of_qj[product_id] += last_size
        self.count_of_j += 1

    def vwap_fast(self, product_id):
        if self.sum_of_qj[product_id] == 0:
            return 0.0
        return self.sum_of_pjqj[product_id] / self.sum_of_qj[product_id]

    def vwap_formated(self, product_id):
        return f"{self.vwap_fast(product_id):010f}"

    def report(self, point_counts=False) -> str:
        buff = ""
        for product_id, product_item in self.prices_n_vols.items():
            buff += f"{product_id}\t{self.vwap_formated(product_id)}"
            if point_counts:
                buff += f"\tpoints:\t{len(product_item)}"
            buff += "\n"
        return buff


src/payload.py
import json
import logging


def split_frames(recv_fn, recv_buffer=4096):
    """
    Read websocket data and parse the
    frames into well formed JSON sentences.
    Once the JSON is loaded it yelds the data to
    the main CoinVWAP listen loop for further processing.
    """
    msg = recv_fn(recv_buffer)
    carret = 0
    while carret < len(msg):
        if len(msg) - carret < 4:
            # Payload too short to make sense, let's read more data
            # TODO: test cover various partial-frame cases!
            msg += recv_fn(recv_buffer)
        # fin, rsv1, rsv2, rsv3, opcode
        # expected fin=1, rsv*=0, opcode=0x1 (OPCODE_TEXT)
        if f"{msg[carret]:08b}" == "10000001":
            length = msg[carret + 2] << 8 | msg[carret + 3]
            content = msg[carret + 4 : carret + length + 4]
            try:
                parsed = json.loads(content)
                yield parsed
                carret = carret + 4 + length
            except json.decoder.JSONDecodeError:
                logging.debug("Buffer was short, reading in more ..")
                msg += recv_fn(recv_buffer)


src/coinvwap.py
import logging
import signal
import socket
import ssl
import sys
from typing import List

from handshake import ProtocolHandler
from payload import split_frames
from vwap import VWAPStore


class Coinvwap:
    """
    Main class handling the connection and websocket listening.
    listen method can be used with threading.
    """

    def __init__(  # pylint: disable=too-many-arguments
        self,
        product_ids: List[str] or None = None,
        url: str = "wss://ws-feed.exchange.coinbase.com/",
        channel: str = "ticker",
        type_field: str = "ticker",
        price_field: str = "price",
        quantity_field: str = "last_size",
    ) -> None:
        self.url = url
        if product_ids:
            self.product_ids = product_ids
        else:
            self.product_ids = ["BTC-USD", "ETH-USD", "ETH-BTC"]
        self.channel = channel
        self.handler = ProtocolHandler(self.url, self.product_ids, self.channel)
        self.vwap = VWAPStore(
            product_ids=self.product_ids,
            price_field=price_field,
            quantity_field=quantity_field,
            type_field=type_field,
        )
        signal.signal(signal.SIGINT, self.signal_handler)
        self.sock = None
        self.connected = False

    def signal_handler(
        self, signal, frame
    ):  # pylint: disable=unused-argument,redefined-outer-name
        self.connected = False
        print("\r-- Interrupted by the user --")
        self.sock.close()
        print("-- Exiting --")
        sys.exit(0)

    def connect(self):
        self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        if self.handler.get_is_secure():
            ssl_context = ssl.create_default_context(purpose=ssl.Purpose.SERVER_AUTH)
            self.sock = ssl_context.wrap_socket(
                self.sock, server_hostname=self.handler.get_host()
            )
        self.sock.connect((self.handler.get_host(), self.handler.get_port()))
        self._switch_protocol()
        self.connected = True

    def _switch_protocol(self):
        # HTTP request to switch protocol to websocket
        logging.debug(self.handler.get_init_request())
        self.sock.send(self.handler.get_init_request())

        # Expected confirmation: 101 Switching Protocols
        data = self.sock.recv(1024)
        logging.debug(data)
        if "HTTP/1.1 101 Switching Protocols" not in str(data):
            # TODO: make a nice custom Exception
            raise Exception("Switching protocols failed!")
        logging.debug("-- Switched to websockets --")

    def listen(self, report_fn=None):
        # subscribe
        logging.debug("-- Channels subscription --")
        self.sock.send(self.handler.get_subscription())

        while self.connected:
            for payload in split_frames(self.sock.recv):
                self.vwap.store(payload)
                if report_fn:
                    report_fn(self.vwap.report(point_counts=True), end="")

    def disconnect(self):
        self.connected = False


src/handshake.py
from __future__ import annotations

import ipaddress
import json
import uuid
from typing import Dict, List
from urllib.parse import urlsplit

from subscribe import ABNF


class ProtocolHandler:  # pylint: disable=too-many-instance-attributes
    """
    Used to handle handshake of the switch from HTTP(S) to websocket.py
    It also wraps the sending of the subscription with Headers.
    """

    def __init__(self, url: str, product_ids: List[str], channel: str) -> None:
        self.url = url
        self.product_ids = product_ids
        self.sec_websocket_key = f"{uuid.uuid4()}=="
        self.switch_headers = self._get_switch_headers()
        self.channel = channel

    def _get_switch_headers_parts(self):
        parsed_url = urlsplit(self.url)
        is_secure = parsed_url.scheme in ["https", "wss"]
        host = parsed_url.hostname
        port = parsed_url.port or (443 if is_secure else 80)
        path = parsed_url.path
        if parsed_url.query:
            path += "?" + parsed_url.query
        return (host, port, is_secure)

    def _get_switch_headers(self) -> Dict:
        self.host, self.port, self.is_secure = self._get_switch_headers_parts()

        self.headers = {}
        self.headers["Host"] = self.build_host(self.host, self.port, self.is_secure)
        self.headers["Upgrade"] = "websocket"
        self.headers["Connection"] = "Upgrade"
        self.headers["Sec-WebSocket-Key"] = self.sec_websocket_key
        self.headers["Sec-WebSocket-Protocol"] = "chat, superchat"
        self.headers["Sec-WebSocket-Version"] = "13"
        return self.headers

    def get_switch_headers(self) -> bytes:
        return str(
            ("".join(f"{key}: {value}\r\n" for key, value in self.headers.items()))
            + "\r\n"
        ).encode()

    def get_host(self):
        return self.host

    def get_port(self):
        return self.port

    def get_is_secure(self):
        return self.is_secure

    def get_init_request(self) -> str:
        init_request = "GET / HTTP/1.1\r\n".encode()
        init_request += self.get_switch_headers()
        return init_request

    def get_subscription(self) -> bytes:
        params = {
            "type": "subscribe",
            "product_ids": self.product_ids,
            "channels": [
                "heartbeat",
                {"name": self.channel, "product_ids": self.product_ids},
            ],
        }
        subscribe = ABNF.create_frame(json.dumps(params), 0x1, fin=1)
        return subscribe.format()

    @staticmethod
    def build_host(host: str, port: int, secure: bool) -> str:
        """
        Build a ``Host`` header.
        """
        # https://www.rfc-editor.org/rfc/rfc3986.html#section-3.2.2
        # IPv6 addresses must be enclosed in brackets.
        try:
            address = ipaddress.ip_address(host)
        except ValueError:
            # host is a hostname
            pass
        else:
            # host is an IP address
            if address.version == 6:
                host = f"[{host}]"

        if port != (443 if secure else 80):
            host = f"{host}:{port}"

        return host


src/subscribe.py
"""
The code below has been heavily borrowed from:
https://github.com/websocket-client/websocket-client/blob/e49e0e88a37c2ba0765eb67522f7ac809fe40709/websocket/_abnf.py#L1

It is responsible for a proper forming the headers and mask
around the subscription request.

----

_abnf.py
websocket - WebSocket client library for Python

Copyright 2022 engn33r

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""

import array
import os
import struct
import sys

native_byteorder = sys.byteorder


def _mask(mask_value, data_value):
    datalen = len(data_value)
    data_value = int.from_bytes(data_value, native_byteorder)
    mask_value = int.from_bytes(
        mask_value * (datalen // 4) + mask_value[: datalen % 4], native_byteorder
    )
    return (data_value ^ mask_value).to_bytes(datalen, native_byteorder)


class ABNF:
    """
    ABNF frame class.
    See http://tools.ietf.org/html/rfc5234
    and http://tools.ietf.org/html/rfc6455#section-5.2
    """

    # operation code values.
    OPCODE_CONT = 0x0
    OPCODE_TEXT = 0x1
    OPCODE_BINARY = 0x2
    OPCODE_CLOSE = 0x8
    OPCODE_PING = 0x9
    OPCODE_PONG = 0xA

    # available operation code value tuple
    OPCODES = (
        OPCODE_CONT,
        OPCODE_TEXT,
        OPCODE_BINARY,
        OPCODE_CLOSE,
        OPCODE_PING,
        OPCODE_PONG,
    )

    # opcode human readable string
    OPCODE_MAP = {
        OPCODE_CONT: "cont",
        OPCODE_TEXT: "text",
        OPCODE_BINARY: "binary",
        OPCODE_CLOSE: "close",
        OPCODE_PING: "ping",
        OPCODE_PONG: "pong",
    }

    # data length threshold.
    LENGTH_7 = 0x7E
    LENGTH_16 = 1 << 16
    LENGTH_63 = 1 << 63

    def __init__(
        self, fin=0, rsv1=0, rsv2=0, rsv3=0, opcode=OPCODE_TEXT, mask=1, data=""
    ):
        """
        Constructor for ABNF. Please check RFC for arguments.
        """
        self.fin = fin
        self.rsv1 = rsv1
        self.rsv2 = rsv2
        self.rsv3 = rsv3
        self.opcode = opcode
        self.mask = mask
        if data is None:
            data = ""
        self.data = data
        self.get_mask_key = os.urandom

    def __str__(self):
        return (
            "fin="
            + str(self.fin)
            + " opcode="
            + str(self.opcode)
            + " data="
            + str(self.data)
        )

    @staticmethod
    def create_frame(data, opcode, fin=1):
        """
        Create frame to send text, binary and other data.
        Parameters
        ----------
        data: <type>
            data to send. This is string value(byte array).
            If opcode is OPCODE_TEXT and this value is unicode,
            data value is converted into unicode string, automatically.
        opcode: <type>
            operation code. please see OPCODE_XXX.
        fin: <type>
            fin flag. if set to 0, create continue fragmentation.
        """
        if opcode == ABNF.OPCODE_TEXT and isinstance(data, str):
            data = data.encode("utf-8")
        # mask must be set if send data from client
        return ABNF(fin, 0, 0, 0, opcode, 1, data)

    def format(self):
        """
        Format this object to string(byte array) to send data to server.
        """
        if any(x not in (0, 1) for x in [self.fin, self.rsv1, self.rsv2, self.rsv3]):
            raise ValueError("not 0 or 1")
        if self.opcode not in ABNF.OPCODES:
            raise ValueError("Invalid OPCODE")
        length = len(self.data)
        if length >= ABNF.LENGTH_63:
            raise ValueError("data is too long")

        # Tip:
        # "{:03b}".format(fin << 7 | 0 << 6 | 0 << 5 | 0 << 4 | 0x1)
        # '10000001'
        frame_header = chr(
            self.fin << 7
            | self.rsv1 << 6
            | self.rsv2 << 5
            | self.rsv3 << 4
            | self.opcode
        ).encode("latin-1")
        if length < ABNF.LENGTH_7:
            frame_header += chr(self.mask << 7 | length).encode("latin-1")
        elif length < ABNF.LENGTH_16:
            frame_header += chr(self.mask << 7 | 0x7E).encode("latin-1")
            frame_header += struct.pack(
                "!H", length
            )  # ! network (= big-endian), H unsigned short
        else:
            frame_header += chr(self.mask << 7 | 0x7F).encode("latin-1")
            frame_header += struct.pack(
                "!Q", length
            )  # ! network (= big-endian), Q unsigned long long

        if not self.mask:
            return frame_header + self.data
        else:
            mask_key = self.get_mask_key(4)
            return frame_header + self._get_masked(mask_key)

    def _get_masked(self, mask_key):
        s = ABNF.mask(mask_key, self.data)

        if isinstance(mask_key, str):
            mask_key = mask_key.encode("utf-8")

        return mask_key + s

    @staticmethod
    def mask(mask_key, data):
        """
        Mask or unmask data. Just do xor for each byte
        Parameters
        ----------
        mask_key: <type>
            4 byte string.
        data: <type>
            data to mask/unmask.
        """
        if data is None:
            data = ""

        if isinstance(mask_key, str):
            mask_key = mask_key.encode("latin-1")

        if isinstance(data, str):
            data = data.encode("latin-1")

        return _mask(array.array("B", mask_key), array.array("B", data))
